src/modules/ekf2/EKF/output_predictor.cpp

```diff
@@ -242,137 +293,258 @@ void OutputPredictor::calculateOutputStates(const uint64_t time_us, const Vector
+	if (_output_new.dt > 0.f) {
+		_output_buffer.push(_output_new);
+		_output_new.dt = 0.0f; // reset time delta to zero for the next accumulation of full rate IMU data
+	}
 
 	// store IMU bias for calculateOutputStates
 	_gyro_bias = gyro_bias;
 	_accel_bias = accel_bias;
 
-	// store the INS states in a ring buffer with the same length and time coordinates as the IMU data buffer
-	_output_buffer.push(_output_new);
-	_output_vert_buffer.push(_output_vert_new);
+	// calculate an average filter update time
+	if ((_time_last_correct_states_us != 0) && (time_delayed_us > _time_last_correct_states_us)) {
+		const float dt = math::constrain((time_delayed_us - _time_last_correct_states_us) * 1e-6f, 0.0001f, 0.03f);
+		_dt_correct_states_avg = 0.8f * _dt_correct_states_avg + 0.2f * dt;
+
+	} else {
+		if (!_output_filter_aligned) {
+			resetQuaternion(time_delayed_us, quat_state);
+			resetHorizontalVelocityTo(time_delayed_us, vel_state.xy());
+			resetVerticalVelocityTo(time_delayed_us, vel_state(2));
+			resetHorizontalPositionTo(time_delayed_us, pos_state.xy());
+			resetVerticalPositionTo(time_delayed_us, pos_state(2));
+		}
+
+		_time_last_correct_states_us = time_delayed_us;
+		return;
+	}
+
+	_time_last_correct_states_us = time_delayed_us;
 
 	// get the oldest INS state data from the ring buffer
 	// this data will be at the EKF fusion time horizon
-	// TODO: there is no guarantee that data is at delayed fusion horizon
-	//       Shouldnt we use pop_first_older_than?
-	const outputSample &output_delayed = _output_buffer.get_oldest();
-	const outputVert &output_vert_delayed = _output_vert_buffer.get_oldest();
+	outputSample output_delayed;
+
+	while (_output_buffer.pop_first_older_than(time_delayed_us, &output_delayed)) {
 
-	// calculate the quaternion delta between the INS and EKF quaternions at the EKF fusion time horizon
-	const Quatf q_error((quat_state.inversed() * output_delayed.quat_nominal).normalized());
+		if (output_delayed.time_us != time_delayed_us) {
+			continue;
+		}
 
-	// convert the quaternion delta to a delta angle
-	const float scalar = (q_error(0) >= 0.0f) ? -2.f : 2.f;
+		if (!_output_filter_aligned) {
 
-	const Vector3f delta_ang_error{scalar * q_error(1), scalar * q_error(2), scalar * q_error(3)};
+			// calculate the quaternion rotation delta from the EKF to output observer states at the EKF fusion time horizon
+			const Quatf q_delta((quat_state * output_delayed.quat_nominal.inversed()).normalized());
+			const Dcmf rot_change{q_delta};
+
+			// add the reset amount to the output observer buffered data
+			for (uint8_t i = 0; i < _output_buffer.get_length(); i++) {
+
+				_output_buffer[i].quat_nominal = (q_delta * _output_buffer[i].quat_nominal).normalized();
+
+				// apply change to velocity
+				_output_buffer[i].vel = rot_change * _output_buffer[i].vel;
+				_output_buffer[i].vel_alternative = rot_change * _output_buffer[i].vel_alternative;
```

Comment:
I see what you're trying to do here, but why correcting the velocities here if they are anyway re-corrected just after?